### PR TITLE
RFC: trying out asyncio and making it do all the work

### PR DESCRIPTION
**Pull request by manannayak**
obviously this requires more work. Just sending this out to hear other
people's thoughts. One big thing missing from this is pausing/resuming
contexts.

### CODE REVIEW

**Comment by LiraNuna** [comment]
It's fantastic to see asynq batching cooperating with asyncio!
I wanted to fork asynq and do this in a few months when we're fully-migrated our data fetches to asynq.

My reasoning was that quora's code base (according to your blog) is still using python2.7 so you will not pursue this kind of change.

This approach makes debugging, error handling and speed much better overall, 👍 to this direction!

File: asynq/tests/test_batching3.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+async def middle2(key):
+    value = await leaf(key)
+    value += '-2b'
+    value = await leaf(value)
+    return value
+
+
+async def leaf(key):
+    result = await ABatchItem(key).get()
```
Comment (LiraNuna) [inline asynq/tests/test_batching3.py]: one thing I am personally pleased about is the fact that `return await` is valid syntax. With asynq, it is possible to use `return (yield ...)` but only with the parens.

File: asynq/batching3.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        self._value = value
+
+    async def get(self):
+        await self.batch.enqueue(self)
+        return self._value
+
+
+class LocalState(threading.local):
+    def __init__(self):
+        super(LocalState, self).__init__()
```
Comment (JelleZijlstra) [inline asynq/batching3.py]: this is Python 3, you can just write `super()` :)

**Comment by LiraNuna** [comment]
What's the status of this? I noticed you are using `asyncio.sleep` for synchronization purposes and I wanted to give you a helping hand with our research.

I made this crude proof of concept that needs cleanup but the concept is sound. We will probably be using it internally once we get it out of poc status: https://gist.github.com/LiraNuna/38a9cff568f9afb34a2d6122218fa61d

**Comment by manannayak** [comment]
interesting. The sleeps are only in tests (to simulate an actual async call to fetch the data) so I'm not too worried about that.

As for the status of this: I'm not sure if we should expect this PR to actually land in the repo since it needs asyncio compatible clients. So it would break the api for existing users of asynq. @JelleZijlstra thoughts?

I'm down to close it as it is.

**Comment by JelleZijlstra** [comment]
We could merge it in, but make `asynq/__init__.py` not import the `py3` subpackage. That way `import asynq` will still work in Python 2.

I guess we could merge this and label it as an experimental extension to asynq. I don't feel strongly about it.

**Comment by LiraNuna** [comment]
We will be switching away from asynq as we would like more control over the event loop that we cannot get with asynq. 

Our batching solution: https://github.com/LiraNuna/aio-batching

**Comment by manannayak** [comment]
sorry, closed it by mistake.

**Comment by LiraNuna** [comment]
Something you might want to know:

We, Patreon, have been using asynq for a while now and are in the process of moving toward using asyncio exclusively. Per our experiments and canaries, **asynq is 10-15% faster than asyncio** but the ability to control the event loop and the interoperability with the various asyncio libraries mean that we could easily regain the speed in the future.

Asynq is a great migration step toward asyncio and we thank you for this library.

**Comment by manannayak** [comment]
Oh that's very interesting, thanks for posting your findings here!

To clarify: in your experiments - did you compare asynq with vanilla asyncio or was there some batching logic implemented on top of asyncio (like this pull request).

**Comment by LiraNuna** [comment]
The experiments were 1:1 regarding batching and produced very similar batching behaviors (our version was slightly better), our implementation could be seen here: https://github.com/LiraNuna/aio-batching

**Comment by SurenNihalani** [comment]
Gonna close this for now since the discussion is older than a few months.

**Pull request closed without merging**